File: silicon/csrc/match.c
// Copyright (c) 2026 Lemon4ksan All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <immintrin.h>
#include <stdint.h>
#include <stddef.h>

// find_match_length_avx2 finds how many bytes at a and b match consecutively (up to max_len).
// Used as the core LZ77 sliding window match finder for Brotli and Deflate compression.
uint64_t find_match_length_avx2(const uint8_t* a, const uint8_t* b, uint64_t max_len) {
    size_t match = 0;

    for (; match + 32 <= max_len; match += 32) {
        __m256i va = _mm256_loadu_si256((const __m256i*)(a + match));
        __m256i vb = _mm256_loadu_si256((const __m256i*)(b + match));

        __m256i cmp = _mm256_cmpeq_epi8(va, vb);
        uint32_t mask = (uint32_t)_mm256_movemask_epi8(cmp);

        if (mask != 0xFFFFFFFF) {
            uint32_t mismatch = (uint32_t)__builtin_ctz(~mask);
            return match + mismatch;
        }
    }

    for (; match < max_len; match++) {
        if (a[match] != b[match]) {
            break;
        }
    }

    return match;
}

File: silicon/simd/match.go
// Copyright (c) 2026 Lemon4ksan All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//go:build amd64 && !purego

package simd

import "unsafe"

// FindMatchLengthVector returns how many consecutive bytes match between a and b (up to maxLen).
// Used as the core LZ77 sliding window match finder for Brotli and Deflate compression.
func FindMatchLengthVector(a, b []byte, maxLen int) int {
	if maxLen <= 0 || len(a) == 0 || len(b) == 0 {
		return 0
	}
	if maxLen > len(a) {
		maxLen = len(a)
	}
	if maxLen > len(b) {
		maxLen = len(b)
	}

	if maxLen >= 32 && hasAVX2 {
		return int(find_match_length_avx2(
			uint64(uintptr(unsafe.Pointer(&a[0]))),
			uint64(uintptr(unsafe.Pointer(&b[0]))),
			uint64(maxLen),
			0, 0, 0,
		))
	}

	for i := 0; i < maxLen; i++ {
		if a[i] != b[i] {
			return i
		}
	}
	return maxLen
}
